keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1187 @@
+  cases.
+- Provide an out-of-tree implementation of this enhancement (using a webhook)
+  that is supported by the Kubernetes org to provide this enhancement
+  functionality to Kubernetes versions where this enhancement is not available.
+- Provide core functionality as a library so that use cases like GitOps,
+  CI/CD pipelines, and auditing can run the same CEL validation checks
+  that the API server does.
+
+### Non-Goals
+
+- Build a comprehensive in-tree policy framework. We believe the ecosystem is
+  best equipped to explore and develop policy frameworks.
+- Full feature parity with mutating admission webhooks. For example, this
+  enhancement is not expected to ever support making requests to external
+  systems.
+- Replace the admission controllers compiled into the API server. 
+- Static or on-initialization specification of admission config. This is a
+  needed feature but should be solved in a general way and not in this KEP
+  (xref: https://github.com/kubernetes/enhancements/issues/1872).
+
+## Proposal
+
+Introduce `MutatingAdmissionPolicy` API, e.g.
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: MutatingAdmissionPolicy
+metadata:
+  name: "sidecar-policy.example.com"
+spec:
+  paramKind:
+    group: mutations.example.com
+    kind: Sidecar
+    version: v1
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE"]
+      resources:   ["pods"]
+  matchConditions:
+    - name: does-not-already-have-sidecar
+      expression: "!object.spec.initContainers.exists(ic, ic.name == params.name)"
+  failurePolicy: Fail
+  mutations:
+    - mutation: >
+        Object{
+          spec: Object.spec{
+            initContainers: [
+              Object.spec.initContainers{
+                name: params.name,
+                image: params.image,
+                args: params.args,
+                restartPolicy: params.restartPolicy
+                // ... other container fields injected here ...
+              }
+            ] + oldObject.spec.initContainers
+          }
+        }
+      mutationValidation:
+        type: Expression
+        expression: "object.spec.initContainers.exists(ic, ic.name == params.name)"
+
+```
+
+The `mutation` field contains a CEL expression that evalutes to a partially
+populated object representing an Server Side Apply "apply configuration". The
+apply configuration is then merged into the request object.
+
+By using Server Side Apply merge algorithms, schema declarations like
+`x-kubernetes-list-type: map`, that control how a merge is performed, will be
+respected.
+
+However, unlike with server side apply, these mutations will not have a field
+manager. This has important implications in how the merge is performed that will
+be discussed in more detail in the below "Unsetting values" section.
+
+The `mutationValidation` field declares how the validate the the mutation has
```

Comment:
Given that the controller knows exactly which fields it's changing and how it should be changed, can this controller do the validation automatically to ensure only the requested fields are updated while others are untouched?

Reply:
Yes, I suggested a possible way to specify this via a `ExactAfterReplay` enum value for `mutationValidation` (https://github.com/kubernetes/enhancements/blob/59ea27c348c7a27c7b233730f6851c4b82605917/keps/sig-api-machinery/3962-mutating-admission-policies/README.md#safety).

I will admit I'm not happy with the enum value name `ExactAfterReplay`, so ideas welcome both on the concept of this validation and of the enum name.